[DCP] Allow users to save and load without creating storage reader and writer

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #117773
* __->__ #117772
* #116248

Right now DCP API requires users to create StorageWriter and StorageReader for every API call. This PR allows users to only pass the checkpointer_id (a path) and use it to read/write a checkpoint without creating a StorageReader and Writer.

Differential Revision: [D52740556](https://our.internmc.facebook.com/intern/diff/D52740556/)

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @LucasLLC

fegin committed:
[DCP][RFC] Allow users to save and load without creating storage reader and writer

TBD

Differential Revision: [D52740556](https://our.internmc.facebook.com/intern/diff/D52740556/)

[ghstack-poisoned]

fegin committed:
Update on "[DCP][RFC] Allow users to save and load without creating storage reader and writer"


Right now DCP API requires users to create StorageWriter and StorageReader for every API call. This PR allows users to only pass the checkpointer_id (a path) and use it to read/write a checkpoint without creating a StorageReader and Writer.

Differential Revision: [D52740556](https://our.internmc.facebook.com/intern/diff/D52740556/)

[ghstack-poisoned]

LucasLLC commented on torch/distributed/checkpoint/state_dict_loader.py:
```diff
@@ -116,6 +120,17 @@ def load(
+                raise RuntimeError(
+                    "`checkpoint_id` must be specificed if storage_reader is None."
+                )
+            # TODO: automatically decide whether to use FSSpecFileSystem
```
I like this idea

LucasLLC commented on torch/distributed/checkpoint/state_dict_saver.py:
```diff
@@ -116,14 +120,19 @@ def save(
-            dumpable_state_dict[key] = (
-                elem.state_dict() if isinstance(elem, Stateful) else elem
-            )
+        if not storage_writer:
```
Nit: could be a good place for a helper function

fegin replied:
I think the `TODO` to decide whether to use `FSSpec` or `FileSystem` should be an util. But this is pretty standalone as StorageWriter and StorageReader are separate classes.

LucasLLC reviewed (commented):
Cool! Couple comments

fegin committed:
Update on "[DCP][RFC] Allow users to save and load without creating storage reader and writer"


Right now DCP API requires users to create StorageWriter and StorageReader for every API call. This PR allows users to only pass the checkpointer_id (a path) and use it to read/write a checkpoint without creating a StorageReader and Writer.

Differential Revision: [D52740556](https://our.internmc.facebook.com/intern/diff/D52740556/)

[ghstack-poisoned]

fegin commented on torch/distributed/checkpoint/state_dict_saver.py:
```diff
@@ -168,13 +178,14 @@ def _async_save(
     executor = ThreadPoolExecutor(max_workers=1)
     f = executor.submit(
-        _save_state_dict,
+        save,
```
It is okay to use `save` because `_stateful_to_state_dict` will be an noop when it is called twice. The reason to call `save` is to ensure that we have the same way to handle the creation of StorageWriter for `async_save` and `save`.

fegin committed:
Update on "[DCP][RFC] Allow users to save and load without creating storage reader and writer"


Right now DCP API requires users to create StorageWriter and StorageReader for every API call. This PR allows users to only pass the checkpointer_id (a path) and use it to read/write a checkpoint without creating a StorageReader and Writer.

Differential Revision: [D52740556](https://our.internmc.facebook.com/intern/diff/D52740556/)

[ghstack-poisoned]

fegin committed:
Update on "[DCP] Allow users to save and load without creating storage reader and writer"


Right now DCP API requires users to create StorageWriter and StorageReader for every API call. This PR allows users to only pass the checkpointer_id (a path) and use it to read/write a checkpoint without creating a StorageReader and Writer.

Differential Revision: [D52740556](https://our.internmc.facebook.com/intern/diff/D52740556/)

[ghstack-poisoned]

fegin committed:
Update on "[DCP] Allow users to save and load without creating storage reader and writer"


Right now DCP API requires users to create StorageWriter and StorageReader for every API call. This PR allows users to only pass the checkpointer_id (a path) and use it to read/write a checkpoint without creating a StorageReader and Writer.

Differential Revision: [D52740556](https://our.internmc.facebook.com/intern/diff/D52740556/)

[ghstack-poisoned]

fegin committed:
Update on "[DCP] Allow users to save and load without creating storage reader and writer"


Right now DCP API requires users to create StorageWriter and StorageReader for every API call. This PR allows users to only pass the checkpointer_id (a path) and use it to read/write a checkpoint without creating a StorageReader and Writer.

Differential Revision: [D52740556](https://our.internmc.facebook.com/intern/diff/D52740556/)

[ghstack-poisoned]

wz337 commented on torch/distributed/checkpoint/filesystem.py:
```diff
@@ -430,14 +435,22 @@ def finish(self, metadata: Metadata, results: List[List[WriteResult]]) -> None:
     def _slice_file(self, file, sinfo: _StorageInfo) -> io.IOBase:
         return _create_file_view(file, sinfo.offset, sinfo.length)
 
+    def reset(self, checkpoint_id: Union[str, os.PathLike, None] = None) -> None:
```
ditto.

wz337 commented on torch/distributed/checkpoint/state_dict_loader.py:
```diff
@@ -26,6 +28,7 @@ def load_state_dict(
         "'load_state_dict' is deprecated and will be removed in future versions. "
         "Please use 'load' instead."
     )
+    storage_reader.reset()
```
Making kwarg checkpoint_id as optional should fix this MYPY complaint?

wz337 reviewed (approved):
Other than the minor comments, overall LGTM!

fegin committed:
Update on "[DCP] Allow users to save and load without creating storage reader and writer"


Right now DCP API requires users to create StorageWriter and StorageReader for every API call. This PR allows users to only pass the checkpointer_id (a path) and use it to read/write a checkpoint without creating a StorageReader and Writer.

Differential Revision: [D52740556](https://our.internmc.facebook.com/intern/diff/D52740556/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 LucasLLC

[ghstack-poisoned]

fegin committed:
Update on "[DCP] Allow users to save and load without creating storage reader and writer"


Right now DCP API requires users to create StorageWriter and StorageReader for every API call. This PR allows users to only pass the checkpointer_id (a path) and use it to read/write a checkpoint without creating a StorageReader and Writer.

Differential Revision: [D52740556](https://our.internmc.facebook.com/intern/diff/D52740556/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 LucasLLC

[ghstack-poisoned]

fegin committed:
Update on "[DCP] Allow users to save and load without creating storage reader and writer"


Right now DCP API requires users to create StorageWriter and StorageReader for every API call. This PR allows users to only pass the checkpointer_id (a path) and use it to read/write a checkpoint without creating a StorageReader and Writer.

Differential Revision: [D52740556](https://our.internmc.facebook.com/intern/diff/D52740556/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 LucasLLC

[ghstack-poisoned]

fegin committed:
Update on "[DCP] Allow users to save and load without creating storage reader and writer"


Right now DCP API requires users to create StorageWriter and StorageReader for every API call. This PR allows users to only pass the checkpointer_id (a path) and use it to read/write a checkpoint without creating a StorageReader and Writer.

Differential Revision: [D52740556](https://our.internmc.facebook.com/intern/diff/D52740556/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 LucasLLC

[ghstack-poisoned]
